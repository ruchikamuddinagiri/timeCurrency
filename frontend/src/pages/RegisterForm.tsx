import { object as zObject, string as zString, TypeOf, z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import { useForm, FormProvider, SubmitHandler } from "react-hook-form";
import { useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import FormInput from "../components/FormInput";
import { Button } from "../components/LoadingButton";
import { api } from "../api/authApi";
import { ApiResponse } from "../api/types";
import useStore from "../store";

const registerSchema = zObject({
  name: zString().min(1, "Full name is required").max(100),
  email: zString()
    .min(1, "Email address is required")
    .email("Email Address is invalid"),
  password: zString()
    .min(1, "Password is required")
    .min(8, "Password must be more than 8 characters")
    .max(32, "Password must be less than 32 characters"),
  passwordConfirm: zString().min(1, "Please confirm your password"),
}).refine((data) => data.password === data.passwordConfirm, {
  path: ["passwordConfirm"],
  message: "Passwords do not match",
});

export type RegisterInput = TypeOf<typeof registerSchema>;

const RegisterPage = () => {
  const appStore = useStore();
  const navigate = useNavigate();

  const methods = useForm<RegisterInput>({
    resolver: zodResolver(registerSchema),
  });

  const {
    reset,
    handleSubmit,
    formState: { isSubmitSuccessful },
  } = methods;

  useEffect(() => {
    if (isSubmitSuccessful) {
      reset();
    }
  }, [isSubmitSuccessful]);

  const registerUser = async (data: RegisterInput) => {
    try {
      appStore.setRequestLoading(true);
      const response = await api.post<ApiResponse>("/auth/register", data);
      appStore.setRequestLoading(false);

      toast.success("User Succesfully registered", {
        position: "top-left",
      });
      // navigate("/emailverification");
    } catch (error: any) {
      appStore.setRequestLoading(false);
      const resMessage =
        (error.response &&
          error.response.data &&
          error.response.data.message) ||
        error.message ||
        error.toString();
      toast.error("User already existed", {
        position: "top-right",
      });
    }
  };

  const onSubmitHandler: SubmitHandler<RegisterInput> = (values) => {
    //console.log("I was called");
    registerUser(values);
  };

  return (
    <section className="py-8 bg-ct-black-600 min-h-screen grid place-items-center">
      <div className="w-full">
        <h1 className="text-4xl xl:text-6xl text-center font-[600] text-ct-black-600 mb-4">
          SIGN UP
        </h1>
        <FormProvider {...methods}>
          <form
            onSubmit={handleSubmit(onSubmitHandler)}
            className="max-w-md w-full mx-auto overflow-hidden shadow-lg bg-ct-dark-200 squared-2x2 p-8 space-y-5"
          >
            <FormInput labelText="Full Name" inputName="name" />
            <FormInput labelText="Email" inputName="email" inputType="email" />
            <FormInput
              labelText="Password"
              inputName="password"
              inputType="password"
            />
            <FormInput
              labelText="Confirm Password"
              inputName="passwordConfirm"
              inputType="password"
            />
            <span className="block">
              Existing User?{" "}
              <Link to="/login" className="text-ct-blue-600">
                Login Here
              </Link>
            </span>
            <Button
              loading={appStore.isLoading}
              text="Sign up"
              color="ct-black-600"
            ></Button>
          </form>
        </FormProvider>
      </div>
    </section>
  );
};
export default RegisterPage;
