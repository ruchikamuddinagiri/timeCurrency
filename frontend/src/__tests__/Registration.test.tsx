import * as React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from "react-router-dom";

import RegisterForm from '../pages/RegisterForm';


jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('Test for Registeration', () => {
  const mockNavigate = jest.fn();
  

  it('renders correctly', () => {

    const { container } = render(<MemoryRouter>
      <RegisterForm />
    </MemoryRouter>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
