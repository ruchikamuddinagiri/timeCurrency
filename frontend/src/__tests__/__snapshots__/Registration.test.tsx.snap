// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent renders correctly 1`] = `
<section
  class="py-8 bg-ct-black-600 min-h-screen grid place-items-center"
>
  <div
    class="w-full"
  >
    <h1
      class="text-4xl xl:text-6xl text-center font-[600] text-ct-black-600 mb-4"
    >
      SIGN UP
    </h1>
    <form
      class="max-w-md w-full mx-auto overflow-hidden shadow-lg bg-ct-dark-200 squared-2x2 p-8 space-y-5"
    >
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="name"
        >
          Full Name
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="name"
          name="name"
          placeholder=""
          type="text"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="email"
        >
          Email
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="email"
          name="email"
          placeholder=""
          type="email"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="password"
          name="password"
          placeholder=""
          type="password"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="passwordConfirm"
        >
          Confirm Password
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="passwordConfirm"
          name="passwordConfirm"
          placeholder=""
          type="password"
        />
      </div>
      <span
        class="block"
      >
        Existing User?
         
        <a
          class="text-ct-blue-600"
          href="/login"
        >
          Login Here
        </a>
      </span>
      <button
        class="w-full py-3 font-semibold rounded-lg outline-none border-none flex justify-center ct-black-600 false"
        type="submit"
      >
        <span>
          Sign up
        </span>
      </button>
    </form>
  </div>
</section>
`;

exports[`Testing rendering for RegisterForm Page renders correctly 1`] = `
<section
  class="py-8 bg-ct-black-600 min-h-screen grid place-items-center"
>
  <div
    class="w-full"
  >
    <h1
      class="text-4xl xl:text-6xl text-center font-[600] text-ct-black-600 mb-4"
    >
      SIGN UP
    </h1>
    <form
      class="max-w-md w-full mx-auto overflow-hidden shadow-lg bg-ct-dark-200 squared-2x2 p-8 space-y-5"
    >
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="name"
        >
          Full Name
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="name"
          name="name"
          placeholder=""
          type="text"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="email"
        >
          Email
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="email"
          name="email"
          placeholder=""
          type="email"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="password"
        >
          Password
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="password"
          name="password"
          placeholder=""
          type="password"
        />
      </div>
      <div
        class="my-3"
      >
        <label
          class="block text-gray-600 mb-1 font-medium"
          for="passwordConfirm"
        >
          Confirm Password
        </label>
        <input
          class="block w-full py-2 px-4 rounded-md bg-gray-100 border-transparent focus:outline-none focus:ring-2 focus:ring-blue-500 focus:bg-white focus:border-transparent"
          id="passwordConfirm"
          name="passwordConfirm"
          placeholder=""
          type="password"
        />
      </div>
      <span
        class="block"
      >
        Existing User?
         
        <a
          class="text-ct-blue-600"
          href="/login"
        >
          Login Here
        </a>
      </span>
      <button
        class="w-full py-3 font-semibold rounded-lg outline-none border-none flex justify-center ct-black-600 false"
        type="submit"
      >
        <span>
          Sign up
        </span>
      </button>
    </form>
  </div>
</section>
`;
